pkg/dogstatsd/parse_events.go

```diff
@@ -0,0 +1,192 @@
+	titleLength int
+	textLength  int
+}
+
+var (
+	eventTimestampPrefix      = []byte("d:")
+	eventHostnamePrefix       = []byte("h:")
+	eventAggregationKeyPrefix = []byte("k:")
+	eventPriorityPrefix       = []byte("p:")
+	eventSourceTypePrefix     = []byte("s:")
+	eventAlertTypePrefix      = []byte("t:")
+	eventTagsPrefix           = []byte("#")
+
+	eventPriorityLow    = []byte("low")
+	eventPriorityNormal = []byte("normal")
+
+	eventAlertTypeError   = []byte("error")
+	eventAlertTypeWarning = []byte("warning")
+	eventAlertTypeInfo    = []byte("info")
+	eventAlertTypeSuccess = []byte("success")
+)
+
+// splitHeaderEvent splits the event and the
+func splitHeaderEvent(message []byte) ([]byte, []byte, error) {
+	sepIndex := bytes.Index(message, colonSeparator)
+	if sepIndex == -1 {
+		return nil, nil, fmt.Errorf("invalid event: %q", message)
+	}
+	return message[:sepIndex], message[sepIndex+1:], nil
+}
+
+func parseHeader(rawHeader []byte) (eventHeader, error) {
+	if len(rawHeader) < 7 {
+		return eventHeader{}, fmt.Errorf("invalid event header: %q", rawHeader)
+	}
+	rawLengths := rawHeader[3 : len(rawHeader)-1]
+	sepIndex := bytes.Index(rawLengths, commaSeparator)
+	if sepIndex == -1 {
+		return eventHeader{}, fmt.Errorf("invalid event header: %q", rawHeader)
+	}
+	rawTitleLength := rawLengths[:sepIndex]
+	rawTextLength := rawLengths[sepIndex+1:]
+	titleLength, err := strconv.ParseInt(string(rawTitleLength), 10, 32)
+	if err != nil {
+		return eventHeader{}, fmt.Errorf("invalid event header: %q", rawHeader)
+	}
+	textLength, err := strconv.ParseInt(string(rawTextLength), 10, 32)
+	if err != nil {
+		return eventHeader{}, fmt.Errorf("invalid event header: %q", rawHeader)
+	}
+	return eventHeader{
+		titleLength: int(titleLength),
+		textLength:  int(textLength),
+	}, nil
+}
+
+func cleanEventText(text []byte) []byte {
+	return bytes.Replace(text, []byte("\\n"), []byte("\n"), -1)
+}
+
+func parseEventTimestamp(rawTimestamp []byte) (int64, error) {
+	return strconv.ParseInt(string(rawTimestamp), 10, 64)
+}
+
+func parseEventPriority(rawPriority []byte) (eventPriority, error) {
+	switch {
+	case bytes.Equal(rawPriority, eventPriorityNormal):
+		return priorityNormal, nil
+	case bytes.Equal(rawPriority, eventPriorityLow):
+		return priorityLow, nil
+	}
+	return priorityNormal, fmt.Errorf("invalid event priority: %q", rawPriority)
+}
+
+func parseEventAlertType(rawAlertType []byte) (alertType, error) {
+	switch {
+	case bytes.Equal(rawAlertType, eventAlertTypeSuccess):
+		return alertTypeSuccess, nil
+	case bytes.Equal(rawAlertType, eventAlertTypeInfo):
+		return alertTypeInfo, nil
+	case bytes.Equal(rawAlertType, eventAlertTypeWarning):
+		return alertTypeWarning, nil
+	case bytes.Equal(rawAlertType, eventAlertTypeError):
+		return alertTypeError, nil
+	}
+	return alertTypeInfo, fmt.Errorf("invalid alert type: %q", rawAlertType)
+}
+
+func applyEventOptionalField(event dogstatsdEvent, optionalField []byte) (dogstatsdEvent, error) {
+	newEvent := event
```

Comment:
Shouldn't we consider providing a pointer to the `dogstatsdEvent` here to modify the data in-place, avoiding a copy? Not sure that events are the most concerned about performances though.

Reply:
IMO we can wait for issues to show up with events and service check before doing that. I could  try replacing by a pointer but it could push the value to the heap depending on how well the escape analysis does it's job in this function.